File: src/net/minecraft/client/render/entity/FishingBobberEntityRenderer.cpp
#include "net/minecraft/client/render/RenderCore.hpp"
#include "net/minecraft/client/render/RenderType.hpp"
#include "net/minecraft/client/render/Tessellator.hpp"
#include "net/minecraft/client/render/entity/EntityRenderDispatcher.hpp"
#include "net/minecraft/client/render/entity/EntityRenderers.hpp"
#include "net/minecraft/entity/Entity.hpp"
#include "net/minecraft/entity/LivingEntity.hpp"
#include "net/minecraft/entity/projectile/FishingBobberEntity.hpp"
#include "net/minecraft/util/math/MathHelper.hpp"
#include "net/minecraft/util/math/Vec3dClient.hpp"
namespace net::minecraft::client::render::entity {
void FishingBobberEntityRenderer::render(
    const net::minecraft::Entity& entity, double x, double y, double z, float yaw, float tickDelta,
    net::minecraft::util::math::MatrixStack& matrices, const net::minecraft::util::math::Matrix4f& projection) {
 (void)yaw;
 const auto* bobber = dynamic_cast<const ::net::minecraft::entity::projectile::FishingBobberEntity*>(&entity);
 if(bobber == nullptr) {
  return;
 }
 beginDraw(matrices, projection);
 // Nothing binds a shader program for the entity stage; a draw submitted with
 // none bound is dropped silently (RenderCore submit). Every entity renderer
 // opens its own pass -- LivingEntityRenderer does, and its scope restores the
 // previous (unbound) program on exit, so inheriting one is never an option.
 const render::RenderPassScope spritePass(render::RenderType::entityCutout());
 matrices.push();
 matrices.translate(static_cast<float>(x), static_cast<float>(y), static_cast<float>(z));
 matrices.scale(0.5f, 0.5f, 0.5f);
 constexpr int atlasCol = 1;
 constexpr int atlasRow = 2;
 bindTexture("/particles.png");
 Tessellator& tessellator = Tessellator::INSTANCE;
 const float uMin = static_cast<float>(atlasCol * 8 + 0) / 128.0f;
 const float uMax = static_cast<float>(atlasCol * 8 + 8) / 128.0f;
 const float vMin = static_cast<float>(atlasRow * 8 + 0) / 128.0f;
 const float vMax = static_cast<float>(atlasRow * 8 + 8) / 128.0f;
 constexpr float quadWidth = 1.0f;
 constexpr float halfWidth = 0.5f;
 constexpr float halfHeight = 0.5f;
 if(dispatcher != nullptr) {
  matrices.rotate(180.0f - dispatcher->yaw_, 0.0f, 1.0f, 0.0f);
  matrices.rotate(-dispatcher->pitch_, 1.0f, 0.0f, 0.0f);
 }
 render::core::setDrawPose(matrices.top());
 tessellator.startQuads();
 tessellator.normal(0.0f, 1.0f, 0.0f);
 tessellator.vertex(0.0f - halfWidth, 0.0f - halfHeight, 0.0, uMin, vMax);
 tessellator.vertex(quadWidth - halfWidth, 0.0f - halfHeight, 0.0, uMax, vMax);
 tessellator.vertex(quadWidth - halfWidth, quadWidth - halfHeight, 0.0, uMax, vMin);
 tessellator.vertex(0.0f - halfWidth, quadWidth - halfHeight, 0.0, uMin, vMin);
 tessellator.draw();
 matrices.pop();
 if(bobber->owner == nullptr || dispatcher == nullptr) {
  endDraw();
  return;
 }
 const net::minecraft::LivingEntity& owner = *bobber->owner;
 float ownerYaw = (owner.prevYaw + (owner.yaw - owner.prevYaw) * tickDelta) * kPiF / 180.0f;
 double sinYaw = MathHelper::sin(ownerYaw);
 double cosYaw = MathHelper::cos(ownerYaw);
 const float swing = owner.getHandSwingProgress(tickDelta);
 const float swingBob = MathHelper::sin(MathHelper::sqrt(swing) * kPiF);
 net::minecraft::util::math::ClientVec3d vec{-0.5, 0.03, 0.8};
 vec.rotateX(-(owner.prevPitch + (owner.pitch - owner.prevPitch) * tickDelta) * kPiF / 180.0f);
 vec.rotateY(-(owner.prevYaw + (owner.yaw - owner.prevYaw) * tickDelta) * kPiF / 180.0f);
 vec.rotateY(swingBob * 0.5f);
 vec.rotateX(-swingBob * 0.7f);
 double rodTipX = owner.prevX + (owner.x - owner.prevX) * static_cast<double>(tickDelta) + vec.x;
 double rodTipY = owner.prevY + (owner.y - owner.prevY) * static_cast<double>(tickDelta) + vec.y;
 double rodTipZ = owner.prevZ + (owner.z - owner.prevZ) * static_cast<double>(tickDelta) + vec.z;
 if(dispatcher->options().thirdPerson) {
  ownerYaw = (owner.lastBodyYaw + (owner.bodyYaw - owner.lastBodyYaw) * tickDelta) * kPiF / 180.0f;
  sinYaw = MathHelper::sin(ownerYaw);
  cosYaw = MathHelper::cos(ownerYaw);
  rodTipX =
      owner.prevX + (owner.x - owner.prevX) * static_cast<double>(tickDelta) - cosYaw * 0.35 - sinYaw * 0.85;
  rodTipY = owner.prevY + (owner.y - owner.prevY) * static_cast<double>(tickDelta) - 0.45;
  rodTipZ =
      owner.prevZ + (owner.z - owner.prevZ) * static_cast<double>(tickDelta) - sinYaw * 0.35 + cosYaw * 0.85;
 }
 const double bobberX = bobber->prevX + (bobber->x - bobber->prevX) * static_cast<double>(tickDelta);
 const double bobberY = bobber->prevY + (bobber->y - bobber->prevY) * static_cast<double>(tickDelta) + 0.25;
 const double bobberZ = bobber->prevZ + (bobber->z - bobber->prevZ) * static_cast<double>(tickDelta);
 const double deltaX = static_cast<float>(rodTipX - bobberX);
 const double deltaY = static_cast<float>(rodTipY - bobberY);
 const double deltaZ = static_cast<float>(rodTipZ - bobberZ);
 render::RenderPassScope passScope(render::RenderType::lines());
 render::core::setEntityColor(0.0f, 0.0f, 0.0f, 0.0f);
 Tessellator::INSTANCE.light(15, 15);
 // The line draw runs on the frame base (the bobber pose popped above).
 render::core::setDrawPose(matrices.top());
 tessellator.start(3); // GL_LINE_STRIP
 tessellator.color(0);
 constexpr int segments = 16;
 for(int i = 0; i <= segments; ++i) {
  const float t = static_cast<float>(i) / static_cast<float>(segments);
  tessellator.vertex(x + deltaX * static_cast<double>(t),
                     y + deltaY * static_cast<double>(t * t + t) * 0.5 + 0.25,
                     z + deltaZ * static_cast<double>(t));
 }
 tessellator.draw();
 endDraw();
}
} // namespace net::minecraft::client::render::entity
#include "net/minecraft/client/entity/EntityClientRendererRegistration.hpp"
#include "net/minecraft/entity/projectile/FishingBobberEntity.hpp"
namespace net::minecraft::entity::projectile {
std::unique_ptr<::net::minecraft::client::render::entity::EntityRenderer>
FishingBobberEntity::ClientRenderer::create() {
 return std::make_unique<::net::minecraft::client::render::entity::FishingBobberEntityRenderer>();
}
} // namespace net::minecraft::entity::projectile
namespace {
static ::net::minecraft::registry::RegisterEntityRenderer<net::minecraft::entity::projectile::FishingBobberEntity>
    autoRendererReg;
} // namespace

File: mcp/src/net/minecraft/client/render/entity/FishingBobberEntityRenderer.java
/*
 * Decompiled with CFR 0.1.1 (FabricMC 57d88659).
 * 
 * Could not load the following classes:
 *  net.fabricmc.api.EnvType
 *  net.fabricmc.api.Environment
 */
package net.minecraft.client.render.entity;

import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.minecraft.client.render.Tessellator;
import net.minecraft.client.render.entity.EntityRenderer;
import net.minecraft.entity.projectile.FishingBobberEntity;
import net.minecraft.util.math.MathHelper;
import net.minecraft.util.math.Vec3d;
import org.lwjgl.opengl.GL11;

@Environment(value=EnvType.CLIENT)
public class FishingBobberEntityRenderer
extends EntityRenderer {
    public void render(FishingBobberEntity arg, double d, double e, double f, float g, float h) {
        GL11.glPushMatrix();
        GL11.glTranslatef((float)d, (float)e, (float)f);
        GL11.glEnable(32826);
        GL11.glScalef(0.5f, 0.5f, 0.5f);
        int n = 1;
        int n2 = 2;
        this.bindTexture("/particles.png");
        Tessellator tessellator = Tessellator.INSTANCE;
        float f2 = (float)(n * 8 + 0) / 128.0f;
        float f3 = (float)(n * 8 + 8) / 128.0f;
        float f4 = (float)(n2 * 8 + 0) / 128.0f;
        float f5 = (float)(n2 * 8 + 8) / 128.0f;
        float f6 = 1.0f;
        float f7 = 0.5f;
        float f8 = 0.5f;
        GL11.glRotatef(180.0f - this.dispatcher.yaw, 0.0f, 1.0f, 0.0f);
        GL11.glRotatef(-this.dispatcher.pitch, 1.0f, 0.0f, 0.0f);
        tessellator.startQuads();
        tessellator.normal(0.0f, 1.0f, 0.0f);
        tessellator.vertex(0.0f - f7, 0.0f - f8, 0.0, f2, f5);
        tessellator.vertex(f6 - f7, 0.0f - f8, 0.0, f3, f5);
        tessellator.vertex(f6 - f7, 1.0f - f8, 0.0, f3, f4);
        tessellator.vertex(0.0f - f7, 1.0f - f8, 0.0, f2, f4);
        tessellator.draw();
        GL11.glDisable(32826);
        GL11.glPopMatrix();
        if (arg.owner != null) {
            float f9 = (arg.owner.prevYaw + (arg.owner.yaw - arg.owner.prevYaw) * h) * (float)Math.PI / 180.0f;
            double d2 = MathHelper.sin(f9);
            double d3 = MathHelper.cos(f9);
            float f10 = arg.owner.getHandSwingProgress(h);
            float f11 = MathHelper.sin(MathHelper.sqrt(f10) * (float)Math.PI);
            Vec3d vec3d = Vec3d.createCached(-0.5, 0.03, 0.8);
            vec3d.rotateX(-(arg.owner.prevPitch + (arg.owner.pitch - arg.owner.prevPitch) * h) * (float)Math.PI / 180.0f);
            vec3d.rotateY(-(arg.owner.prevYaw + (arg.owner.yaw - arg.owner.prevYaw) * h) * (float)Math.PI / 180.0f);
            vec3d.rotateY(f11 * 0.5f);
            vec3d.rotateX(-f11 * 0.7f);
            double d4 = arg.owner.prevX + (arg.owner.x - arg.owner.prevX) * (double)h + vec3d.x;
            double d5 = arg.owner.prevY + (arg.owner.y - arg.owner.prevY) * (double)h + vec3d.y;
            double d6 = arg.owner.prevZ + (arg.owner.z - arg.owner.prevZ) * (double)h + vec3d.z;
            if (this.dispatcher.options.thirdPerson) {
                f9 = (arg.owner.lastBodyYaw + (arg.owner.bodyYaw - arg.owner.lastBodyYaw) * h) * (float)Math.PI / 180.0f;
                d2 = MathHelper.sin(f9);
                d3 = MathHelper.cos(f9);
                d4 = arg.owner.prevX + (arg.owner.x - arg.owner.prevX) * (double)h - d3 * 0.35 - d2 * 0.85;
                d5 = arg.owner.prevY + (arg.owner.y - arg.owner.prevY) * (double)h - 0.45;
                d6 = arg.owner.prevZ + (arg.owner.z - arg.owner.prevZ) * (double)h - d2 * 0.35 + d3 * 0.85;
            }
            double d7 = arg.prevX + (arg.x - arg.prevX) * (double)h;
            double d8 = arg.prevY + (arg.y - arg.prevY) * (double)h + 0.25;
            double d9 = arg.prevZ + (arg.z - arg.prevZ) * (double)h;
            double d10 = (float)(d4 - d7);
            double d11 = (float)(d5 - d8);
            double d12 = (float)(d6 - d9);
            GL11.glDisable(3553);
            GL11.glDisable(2896);
            tessellator.start(GL11.GL_LINE_STRIP);
            tessellator.color(0);
            int n3 = 16;
            for (int i = 0; i <= n3; ++i) {
                float f12 = (float)i / (float)n3;
                tessellator.vertex(d + d10 * (double)f12, e + d11 * (double)(f12 * f12 + f12) * 0.5 + 0.25, f + d12 * (double)f12);
            }
            tessellator.draw();
            GL11.glEnable(2896);
            GL11.glEnable(3553);
        }
    }
}

